clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+
+        AddPartitionsToTransactionCallback() {
+            super(null);
+        }
+
+        @Override
+        public void handleResponse(AbstractResponse response) {
+            AddPartitionsToTxnResponse addPartitionsToTxnResponse = (AddPartitionsToTxnResponse) response;
+            Errors error = addPartitionsToTxnResponse.error();
+            if (error == Errors.NONE) {
+                partitionsInTransaction.addAll(pendingPartitionsToBeAddedToTransaction);
+                pendingPartitionsToBeAddedToTransaction.clear();
+            } else if (error == Errors.COORDINATOR_NOT_AVAILABLE || error == Errors.NOT_COORDINATOR) {
+                needsCoordinator(FindCoordinatorRequest.CoordinatorType.TRANSACTION, transactionalId);
+                reenqueue();
+            } else if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS) {
+                reenqueue();
+            } else if (error == Errors.INVALID_PID_MAPPING || error == Errors.INVALID_TXN_STATE) {
+                throw error.exception();
+            } else if (error == Errors.INVALID_PRODUCER_EPOCH) {
+                isFenced = true;
+                throw error.exception();
+            } else if (error == Errors.TOPIC_AUTHORIZATION_FAILED) {
+                throw error.exception();
+            } else {
+                throw Errors.UNKNOWN.exception();
+            }
+        }
+
+        @Override
+        public void reenqueue() {
+            pendingTransactionalRequests.add(addPartitionsToTransactionRequest(true));
+        }
+    }
+
+    private class FindCoordinatorCallback extends TransactionalRequestCallBack {
+        private final FindCoordinatorRequest.CoordinatorType type;
+        private final String coordinatorKey;
+
+        FindCoordinatorCallback(FindCoordinatorRequest.CoordinatorType type, String coordinatorKey) {
+            super(null);
+            this.type = type;
+            this.coordinatorKey = coordinatorKey;
+        }
+        @Override
+        public void handleResponse(AbstractResponse responseBody) {
+            FindCoordinatorResponse findCoordinatorResponse = (FindCoordinatorResponse) responseBody;
+            if (findCoordinatorResponse.error() == Errors.NONE) {
+                Node node = findCoordinatorResponse.node();
+                switch (type) {
+                    case GROUP:
+                        consumerGroupCoordinator = node;
+                        break;
+                    case TRANSACTION:
+                        transactionCoordinator = node;
+                }
+            } else if (findCoordinatorResponse.error() == Errors.COORDINATOR_NOT_AVAILABLE) {
+                reenqueue();
+            } else if (findCoordinatorResponse.error() == Errors.GROUP_AUTHORIZATION_FAILED) {
+                throw Errors.GROUP_AUTHORIZATION_FAILED.exception();
+            } else {
+                throw Errors.UNKNOWN.exception();
```

Comment:
We're losing the error code from the response, which will definitely be useful if we ever happen to hit this case. In the consumer, we tend to do something like this:
```java
throw new KafkaException("Unexpected error in response: " + response.error().message());
```